### PR TITLE
RFC : Attempt at caching the lindblad operator construct, WIP : Docs

### PR DESCRIPTION
**Pull request by amitjamadagni**
Flag has been passed through `QuEquation` and this has been used to cache the lindblad operator construct. Also for `MCWF` construct uncached version has been used as we do not need the lindblad operator construct.

### CODE REVIEW

**Comment by amitjamadagni** [comment]
@acroy the build fails on nightly version, the following is the error

```
WARNING: both QuBase and Base export "normalize!"; uses of it in module Main must be qualified
ERROR: LoadError: LoadError: UndefVarError: normalize! not defined
```

I guess `normalize!` is defined on the latest Base but not on v0.4, if that is the case will a `Base.normalize!` in `QuBase.jl` fix the error (I was of the thinking the tests might fail on v0.4 if this is done ?!).

**Comment by amitjamadagni** [comment]
@acroy the commits dont seem to work on v0.3.7 they give the following error, but they do build on v 0.4 as projected by travis. Here is the error on v0.3.7

``` julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.3.7 (2015-03-23 21:36 UTC)
 _/ |\__'_|_|_|\__'_|  |  
|__/                   |  x86_64-linux-gnu

julia> include("QuDynamics.jl/src/QuDynamics.jl")
ERROR: type: instantiate_type: expected TypeConstructor, got Function
 in include at ./boot.jl:245
 in include_from_node1 at ./loading.jl:128
 in include at ./boot.jl:245
 in include_from_node1 at ./loading.jl:128
while loading QuDynamics.jl/src/quequations.jl, in expression starting on line 189
while loading QuDynamics.jl/src/QuDynamics.jl, in expression starting on line 8
```

Also a review on the docs and tests would be helpful. I guess this is ready except for the error that it wont run on v0.3.7.

File: src/propmachinery.jl [hunk 1]
```diff
@@ -47,7 +47,7 @@ QuStateEvolution{QPM<:QuPropagatorMethod, QM<:QuBase.AbstractQuMatrix}(eq::QuLin
 
 QuStateEvolution{QPM<:QuPropagatorMethod, QM<:QuBase.AbstractQuMatrix, COT<:QuBase.AbstractQuMatrix}(hamiltonian::QuBase.AbstractQuMatrix, collapse_ops::Vector{COT}, init_state::QM, tlist, method::QPM) = QuStateEvolution{QPM,QM,QuLindbladMasterEq}(QuLindbladMasterEq(hamiltonian,collapse_ops), init_state, tlist, method)
 
-QuStateEvolution{QPM<:QuPropagatorMethod, QV<:QuBase.AbstractQuVector, COT<:QuBase.AbstractQuMatrix}(hamiltonian::QuBase.AbstractQuMatrix, collapse_ops::Vector{COT}, init_state::QV, tlist, method::QPM) = QuStateEvolution{QPM,QV,QuLindbladMasterEq}(QuLindbladMasterEq(hamiltonian,collapse_ops), init_state, tlist, method)
+QuStateEvolution{QPM<:QuPropagatorMethod, QV<:QuBase.AbstractQuVector, COT<:QuBase.AbstractQuMatrix}(hamiltonian::QuBase.AbstractQuMatrix, collapse_ops::Vector{COT}, init_state::QV, tlist, method::QPM) = QuStateEvolution{QPM,QV,QuLindbladMasterEq}(QuLindbladMasterEqUncached(hamiltonian,collapse_ops), init_state, tlist, method)
```
Comment (acroy) [inline src/propmachinery.jl:51]: Here we should have a comment saying that this constructor applies to Lindblad equations which are solved by a wave-function method (e.g. MCWF) and hence the Lindblad matrix is not cached.

**Comment by acroy** [comment]
You need to use `@compat Union{...}` because in 0.3 it used to be `Union(...)`.

Could you also add the caching flag to the other equation types? I guess `<: QuEquation{1}` should be enough for now.

Did you have time to run the benchmarks?

**Comment by amitjamadagni** [comment]
@acroy done with the changes ! Benchmarks are still a WIP both for MCWF and also the lindblad op issue. I hope to have it by the next meeting ! Please do let me know if this is ready for a merge ?!

File: src/quequations.jl [hunk 2]
```diff
@@ -1,4 +1,4 @@
-abstract QuEquation
+abstract QuEquation{CF}
```
Comment (acroy) [inline src/quequations.jl:4]: Please add a comment to explain what `CF` means.

**Comment by acroy** [comment]
I would wait with merging until we have the benchmarks - maybe there is something else to fix ...

**Comment by amitjamadagni** [comment]
@acroy  I have preformed some benchmarks and here are the [results](https://github.com/amitjamadagni/Notes-Notebooks/blob/master/Notebooks/Jaynes%20Cummings%20MCWF%20Benchmarks.ipynb).

**Comment by amitjamadagni** [comment]
@acroy the benchmarks on the current version of the master and after the changes reflected in the current PR are as follows : 

Current Master : https://github.com/amitjamadagni/Notes-Notebooks/blob/master/Notebooks/JC_Model_uncahe.ipynb

After changes :  https://github.com/amitjamadagni/Notes-Notebooks/blob/master/Notebooks/Jaynes%20Cummings%20MCWF%20Benchmarks.ipynb

There is a good amount of speed up after the changes. I guess this is ready to go !

**Comment by acroy** [comment]
Nice speed-up! Could you modify the benchmark such, that you have for a given `N` a set of `ntraj`, e.g. for `N=10` you test `ntraj=200,300,400,500`? For this particular case you can always use the RWA Hamiltonian. Basically, the computational cost should scale linearly with `ntraj` and also with `N`, which would be nice to see.

**Comment by amitjamadagni** [comment]
@acroy I guess it does increase linearly, here are the results : 
https://github.com/amitjamadagni/Notes-Notebooks/blob/master/Notebooks/MCWF_Cache_N_vs_ntraj.ipynb

**Comment by acroy** [comment]
The results seem to be missing in that notebook ...

**Comment by amitjamadagni** [comment]
It seems I forgot to save the notebook before committing ... I have updated it but if it does not show it is recorded here in the blame at the end !

https://github.com/amitjamadagni/Notes-Notebooks/blame/master/Notebooks/MCWF_Cache_N_vs_ntraj.ipynb

**Comment by acroy** [comment]
Thanks. The behavior is reasonably linear in the considered range:
image

If you would plot the data for the uncached version, the slope should be (much) larger ...

I guess this PR is ready. Maybe you could add the benchmark to the repo (since we didn't have one for MCWF) and in the future we can tune it to produce those kind of plots automatically.

**Comment by amitjamadagni** [comment]
@acroy the benchmark is already added to perf. Do we also add the latest notebook, that is the fixed N vs traj plot to the perf (I guess it is just changing the states in the benchmark and not much different from a general one) ?!

**Comment by acroy** [comment]
The other one should be sufficient. As you say it's just a special set of configurations ...

**Comment by amitjamadagni** [comment]
Squashing and merging this !

**Comment by amitjamadagni** [comment]
@acroy the tests seem to fail on the release version. I guess it is because of the random index generation in the tests. Though it passes on the mac build it fails on the linux build. I guess the tests should pass but then any insight on this would be helpful ?!

**Comment by acroy** [comment]
`rhos` has only `length(tlist)-1` elements and by chance you hit `tlist[end]` in the Linux test. I would use `ind=rand(1:(length(tlist)-1))` instead, which should be safe.

**Comment by amitjamadagni** [comment]
Tests pass ! Merging this !

**Pull request merged**